src/ray/raylet/worker_pool.cc

```diff
@@ -1269,8 +1258,15 @@ void WorkerPool::PrestartWorkers(const TaskSpecification &task_spec, int64_t bac
 
 bool WorkerPool::DisconnectWorker(const std::shared_ptr<WorkerInterface> &worker,
                                   rpc::WorkerExitType disconnect_type) {
-  runtime_env_manager_.RemoveURIReference(worker->WorkerId().Hex());
   auto &state = GetStateForLanguage(worker->GetLanguage());
+  auto it = state.worker_processes.find(worker->GetStartupToken());
+  if (it != state.worker_processes.end()) {
+    it->second.num_registered_workers--;
+    if (it->second.num_registered_workers == 0 && it->second.num_starting_workers == 0) {
+      DeleteRuntimeEnvIfNeeded(it->second.runtime_env_info.serialized_runtime_env());
```

Comment:
For this API, my impression is the right name is 

```
DeleteRuntimeEnvIfPossible
```

or 
```
MarkRuntimeEnvUnused
```
Wdyt?

Reply:
`CreateRuntimeEnvOrGet` and `DeleteRuntimeEnvIfPossible`? I like symmetrical names. 😄

Reply:
Hmm imo these two has a slightly different semantics to have symmetric names (create is get Or create , and delete it Delete or do nothing). But it is up to you.